### DISCUSSION TITLE
using exec() inside uasyncio task

### DISCUSSION THREAD

**Discussion by rootfrogsMakes** [micropython/micropython#11800] [Using MicroPython]
Hello,
I am struggeling a little bit with the exec command calling inside of a task.

Since my code is very large, I have reduced it here to the core of my problem:

```python
async def mainLoop():
    while True:
        await startFlag.wait()
        print("before exec")
        exec("uasyncio.sleep(1)")
        print("after exec")
```

The `mainLoop` was started with `uasyncio.create_task(mainLoop())`, so it is running as a task.
In this code, the exec call returns immediately without waiting any time. 
What I need is: exec should return if the given time is elapsed (same behavior, like i directly call `await uasyncio.sleep(1)` without exec)

I tried it also with:

```python
exec("time.sleep(1)")             #this will block all running tasks for 1 second
exec("uasyncio.sleep(1)")         #this returns immediately without waiting any time
exec("await uasyncio.sleep(1)")   #this leads to the error: 'await' outside function
```

I need to call the code here with exec. Using eval is not possible at this point because in the original code there is a function call inside exec that does some additional function calls.

Does anyone have an idea how to get a sleep function by using an exec() call, which does not block all running tasks and only returns after time has elapsed?

**Comment by bixb922** [comment]
asyncio.sleep() needs await. But even with that, exec("await uasyncio.sleep(1)") will not work because it throws a Syntax Error "await outside function" error. CPython also does.

**Comment by peterhinch** [comment]
I'm not sure if it's possible to execute dynamically generated `asyncio` code from an `asyncio` program. You want to avoid creating two `asyncio` instances: there needs to be only one event loop. To have any chance of working I think you'd need to pass the event loop to `exec` and have the generated code call event loop methods (e.g. `loop.create_task()`).

Will this work (in CPython or MP)? I really don't know...

**Comment by peterhinch** [comment]
[EDIT] Removed stupid bug.
OK, I was sufficiently interested in this to try it.
tl;dr it works.
```python
code = '''
import uasyncio as asyncio
async def test():
    for n in range(10):
        await asyncio.sleep_ms(500)
        print(n)
async def run():
    print("Run started")
    tsk = loop.create_task(test())
    print("Created task")
    await tsk
    done.set()
asyncio.create_task(run())
'''
import uasyncio as asyncio
async def twiddle():
    while True:
        print("Waiting")
        await asyncio.sleep_ms(600)

async def main():
    evt = asyncio.Event()
    globs = {'done': evt, 'loop': asyncio.get_event_loop()}
    exec(code, globs)
    tw = asyncio.create_task(twiddle())
    await evt.wait()
    tw.cancel()
    print("Done")

asyncio.run(main())
```
This script works under the Unix build.

**Reply by rootfrogsMakes** [reply]
Thanks for the reply.
Nice idea, I will evaluate it, if this also works with my original complete code.
